File: battle_engine/app/contracts/consumers/get_pokemon_template_moves_contract.rb
# frozen_string_literal: true

module Contracts
  module Consumers
    class GetPokemonTemplateMovesContract < Dry::Validation::Contract
      params do
        required(:player_id).filled(:integer)
        required(:pokemon_template_id).filled(:integer)
        optional(:timestamp).maybe(:string)
      end
    end
  end
end

File: battle_real_time/lib/battle_real_time/contracts/publishers/player_actions/get_pokemon_template_moves_contract.ex
defmodule BattleRealTime.Contracts.Publishers.PlayerActions.GetPokemonTemplateMovesContract do
  @moduledoc """
  Validates the payload for the get_pokemon_template_moves player action, under the player_actions queue.
  """
  use BattleRealTime.Contracts.Contract

  @primary_key false
  embedded_schema do
    field(:player_id, :string)
    field(:pokemon_template_id, :integer)
    field(:timestamp, :string)
  end

  def changeset(struct, params) do
    struct
    |> cast(params, [:player_id, :pokemon_template_id, :timestamp])
    |> validate_required([:player_id, :pokemon_template_id])
  end
end
